### PR TITLE
ARROW-9810: [C++] Generalized nested reconstruction helpers

### PR DESCRIPTION
**Pull request by emkornfield**
This adds helper methods for reconstructing all necessary metadata
for arrow types.  For now this doesn't handle null_slot_usage (i.e.
children of FixedSizeList), it throws exceptions when nulls are
encountered in this case.

The unit tests demonstrate how to use the helper methods in combination
with LevelInfo (generated from parquet/arrow/schema.h) to reconstruct
the metadata.

- Refactors necessary APIs to use LevelInfo and makes use of them in
  column_reader
- Adds implementation for reconstructing list validity bitmaps using rep/def levels..
- Adds implementation for reconstruction list lengths (for rep/dev levels)
- Adds dynamic dispatch for level comparison algorithms for AVX2  and BMI2.
- Adds a pextract alternative that uses BitRunReader that can be
  used as a fallback.

### CODE REVIEW

File: cpp/src/parquet/level_comparison.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                               Predicate predicate) {
+  // Both clang and GCC can vectorize this automatically with SSE4/AVX2.
+  uint64_t mask = 0;
+  for (int x = 0; x < num_levels; x++) {
+    mask |= static_cast<uint64_t>(predicate(levels[x]) ? 1 : 0) << x;
+  }
+  return ::arrow::BitUtil::ToLittleEndian(mask);
+}
+
+/// Builds a  bitmap where each set bit indicates the corresponding level is greater
```
Comment (pitrou) [inline cpp/src/parquet/level_comparison.h:52]: I don't understand why you're doing this. Integer comparisons are cheap, while writing and reading bitmaps is expensive. While go through an intermediate bitmap (and then spend some time trying to optimize it?).

**Comment by pitrou** [comment]
Hmm... I don't know if this is easy to tear apart, but I'd appreciate if the PR concentrated on the functionality, and micro-optimizations (SIMD or otherwise) separated into later PRs. Would that be reasonably doable?

**Comment by emkornfield** [comment]
> Hmm... I don't know if this is easy to tear apart, but I'd appreciate if the PR concentrated on the functionality, and micro-optimizations (SIMD, BMI or otherwise) separated into later PRs. Would that be reasonably doable?

How about I remove the net new bitmap related code?  The SIMD/BMI code was already present hidden behind compiler flags instead of runtime dispatched.  The runtime dispatch is cleaning up technical debt to allow users to make use of these in prepackaged binaries (the [PR](https://github.com/apache/arrow/pull/6985) that introduced them showed a 20% improvement on our end to end parquet->Arrow reading benchmarks).

**Comment by emkornfield** [comment]
@pitrou I've removed the code for reconstructing list information based on bitmaps, I would appreciate keeping the rest of the in this PR.

**Comment by emkornfield** [comment]
Closing in favor of: https://github.com/apache/arrow/pull/8177

**Pull request closed without merging**